modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -1807,17 +1807,22 @@ public CodegenModel fromModel(String name, Schema schema) {
                 m.allVars = new ArrayList<CodegenProperty>();
                 if (composed.getAllOf() != null) {
                     int modelImplCnt = 0; // only one inline object allowed in a ComposedModel
+                    int modelDiscriminators = 0; // only one discriminator allowed in a ComposedModel
                     for (Schema innerSchema : composed.getAllOf()) { // TOOD need to work with anyOf, oneOf as well
-                        if (m.discriminator == null) {
+                        if (m.discriminator == null && innerSchema.getDiscriminator() != null) {
                             LOGGER.debug("discriminator is set to null (not correctly set earlier): {}", name);
-                            m.discriminator = createDiscriminator(name, schema);
+                            m.discriminator = createDiscriminator(name, innerSchema);
```

Comment:
Per here: https://swagger.io/specification/#discriminatorObject
We should be able to set the discriminator on the parent composed schema, but that's not yet working because of this bug in swagger parser: https://github.com/swagger-api/swagger-parser/issues/1269
In place of having discriminator at the root level of a composed schema, we should be able to use it from our definition in our allOf schemas, so I updated this code to set the discriminator on self if one of its ancestor allOf schemas define discriminator.
Please see my ParentPet class which demonstrates this usage at: https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/modules/openapi-generator/src/test/resources/2_0/python-client-experimental/petstore-with-fake-endpoints-models-for-testing.yaml#L2072

NOTE: the original invocation here `m.discriminator = createDiscriminator(name, schema);` did not do anything because we already ran that exact line above in line 1772.

Reply:
I think this seems reasonable, but I'd like to have someone else on the @OpenAPITools/generator-core-team review as well since it affects all generators.